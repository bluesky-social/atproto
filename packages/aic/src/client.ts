import { DidableKey } from '@adxp/crypto'
import axios from 'axios'
import { Document, hashAndFindDid } from './document'
import { Create, create, Operation, UpdateOperation } from './types'
import * as cbor from '@ipld/dag-cbor'
import * as uint8arrays from 'uint8arrays'
import { check, cidForData } from '@adxp/common'

export class AicClient {
  constructor(public url: string) {}

  async getDocument(did: string): Promise<Document> {
    const res = await axios.get(`${this.url}/${did}`)
    return res.data
  }

  async getOperationLog(did: string): Promise<Operation[]> {
    const res = await axios.get(`${this.url}/log/${did}`)
    return res.data.log
  }

  async createDid(
    signingKey: DidableKey,
    recoveryKey: string,
    username: string,
    service: string,
  ): Promise<string> {
    const op: Partial<Create> = {
      type: 'create',
      signingKey: signingKey.did(),
      recoveryKey,
      username,
      service,
      prev: null,
    }
    const signedOp = await this.signOperation(op, signingKey)
    if (!check.is(signedOp, create)) {
      throw new Error('Not a valid create operation')
    }
    const did = await hashAndFindDid(signedOp)
    await axios.post(`${this.url}/${did}`, signedOp)
    return did
  }

  async makeUpdate(
    did: string,
    op: Partial<UpdateOperation>,
    signingKey: DidableKey,
  ) {
    const log = await this.getOperationLog(did)
    if (log.length === 0) {
      throw new Error(`Could not make update: DID does not exist: ${did}`)
    }
    const prev = await cidForData(log[log.length - 1])
    const signedOp = this.signOperation(
      {
        ...op,
        prev: prev.toString(),
      },
      signingKey,
    )
    await axios.post(`${this.url}/${did}`, signedOp)
  }

  async rotateSigningKey(did: string, newKey: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'rotate_signing_key', key: newKey },
      signingKey,
    )
  }

  async rotateRecoveryKey(did: string, newKey: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'rotate_recovery_key', key: newKey },
      signingKey,
    )
  }

  async udpateUsername(did: string, username: string, signingKey: DidableKey) {
    return this.makeUpdate(
      did,
      { type: 'update_username', username },
      signingKey,
    )
  }

  async updateService(did: string, service: string, signingKey: DidableKey) {
    return this.makeUpdate(did, { type: 'update_service', service }, signingKey)
  }
}

export default AicClient
