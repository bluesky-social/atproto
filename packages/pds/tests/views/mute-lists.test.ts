import AtpAgent, { AtUri } from '@atproto/api'
import { runTestServer, CloseFn, TestServerInfo, forSnapshot } from '../_util'
import { SeedClient } from '../seeds/client'
import basicSeed from '../seeds/basic'
import { RecordRef } from '../seeds/client'

describe('pds views with mutes from mute lists', () => {
  let server: TestServerInfo
  let agent: AtpAgent
  let close: CloseFn
  let sc: SeedClient

  let alice: string
  let bob: string
  let carol: string
  let dan: string

  beforeAll(async () => {
    server = await runTestServer({
      dbPostgresSchema: 'views_mute_lists',
    })
    close = server.close
    agent = new AtpAgent({ service: server.url })
    sc = new SeedClient(agent)
    await basicSeed(sc)
    alice = sc.dids.alice
    bob = sc.dids.bob
    carol = sc.dids.carol
    dan = sc.dids.dan
    // add follows to ensure mutes work even w follows
    await sc.follow(carol, dan)
    await sc.follow(dan, carol)
    await server.processAll()
  })

  afterAll(async () => {
    await close()
  })

  let listUri: string
  let listCid: string

  it('creates a list with some items', async () => {
    const avatar = await sc.uploadFile(
      alice,
      'tests/image/fixtures/key-portrait-small.jpg',
      'image/jpeg',
    )
    // alice creates mute list with bob & carol that dan uses
    const list = await agent.api.app.bsky.graph.list.create(
      { repo: alice },
      {
        name: 'alice mutes',
        purpose: 'app.bsky.graph.defs#modlist',
        description: 'big list of mutes',
        avatar: avatar.image,
        createdAt: new Date().toISOString(),
      },
      sc.getHeaders(alice),
    )
    listUri = list.uri
    listCid = list.cid
    await agent.api.app.bsky.graph.listitem.create(
      { repo: alice },
      {
        subject: sc.dids.bob,
        list: list.uri,
        reason: 'because',
        createdAt: new Date().toISOString(),
      },
      sc.getHeaders(alice),
    )
    await agent.api.app.bsky.graph.listitem.create(
      { repo: alice },
      {
        subject: sc.dids.carol,
        list: list.uri,
        reason: 'idk',
        createdAt: new Date().toISOString(),
      },
      sc.getHeaders(alice),
    )
  })

  it('uses a list for mutes', async () => {
    await agent.api.app.bsky.graph.muteActorList(
      {
        list: listUri,
      },
      { encoding: 'application/json', headers: sc.getHeaders(dan) },
    )
  })

  it('flags mutes in threads', async () => {
    const res = await agent.api.app.bsky.feed.getPostThread(
      { depth: 1, uri: sc.posts[alice][1].ref.uriStr },
      { headers: sc.getHeaders(dan) },
    )
    expect(forSnapshot(res.data.thread)).toMatchSnapshot()
  })

  it('does not show reposted content from a muted account in  author feed', async () => {
    await sc.repost(alice, sc.posts[carol][0].ref)

    const res = await agent.api.app.bsky.feed.getAuthorFeed(
      { actor: alice },
      { headers: sc.getHeaders(dan) },
    )
    expect(
      res.data.feed.some((post) => [bob, carol].includes(post.post.author.did)),
    ).toBe(false)
  })

  it('removes content from muted users on getTimeline', async () => {
    const res = await agent.api.app.bsky.feed.getTimeline(
      { limit: 100 },
      { headers: sc.getHeaders(dan) },
    )
    expect(
      res.data.feed.some((post) => [bob, carol].includes(post.post.author.did)),
    ).toBe(false)
  })

  it('flags muted users on getPopular', async () => {
    for (let i = 0; i < 15; i++) {
      const name = `user${i}`
      await sc.createAccount(name, {
        handle: `user${i}.test`,
        email: `user${i}@test.com`,
        password: 'password',
      })
      await sc.like(sc.dids[name], sc.posts[alice][0].ref)
      await sc.like(sc.dids[name], sc.posts[bob][0].ref)
      await sc.like(sc.dids[name], sc.posts[carol][0].ref)
      await sc.like(sc.dids[name], sc.posts[dan][0].ref)
    }

    const res = await agent.api.app.bsky.unspecced.getPopular(
      {},
      { headers: sc.getHeaders(dan) },
    )
    expect(
      res.data.feed.some((post) => [bob, carol].includes(post.post.author.did)),
    ).toBe(false)
  })

  it('returns mute status on getProfile', async () => {
    const res = await agent.api.app.bsky.actor.getProfile(
      { actor: carol },
      { headers: sc.getHeaders(dan) },
    )
    expect(res.data.viewer?.muted).toBe(true)
    expect(res.data.viewer?.mutedByList?.uri).toBe(listUri)
  })

  it('returns mute status on getProfiles', async () => {
    const res = await agent.api.app.bsky.actor.getProfiles(
      { actors: [alice, carol] },
      { headers: sc.getHeaders(dan) },
    )
    expect(res.data.profiles[0].viewer?.muted).toBe(false)
    expect(res.data.profiles[0].viewer?.mutedByList).toBeUndefined()
    expect(res.data.profiles[1].viewer?.muted).toBe(true)
    expect(res.data.profiles[1].viewer?.mutedByList?.uri).toEqual(listUri)
  })

  it('does not return notifs for muted accounts', async () => {
    const res = await agent.api.app.bsky.notification.listNotifications(
      {
        limit: 100,
      },
      { headers: sc.getHeaders(dan) },
    )
    expect(
      res.data.notifications.some((notif) =>
        [bob, carol].includes(notif.author.did),
      ),
    ).toBeFalsy()
  })

  it('flags muted accounts in in get suggestions', async () => {
    // unfollow so they _would_ show up in suggestions if not for mute
    await sc.unfollow(dan, carol)

    const res = await agent.api.app.bsky.actor.getSuggestions(
      {
        limit: 100,
      },
      { headers: sc.getHeaders(dan) },
    )
    for (const actor of res.data.actors) {
      if ([bob, carol].includes(actor.did)) {
        expect(actor.viewer?.muted).toBe(true)
        expect(actor.viewer?.mutedByList?.uri).toEqual(listUri)
      } else {
        expect(actor.viewer?.muted).toBe(false)
        expect(actor.viewer?.mutedByList).toBeUndefined()
      }
    }
  })

  it('returns the contents of a list', async () => {
    const res = await agent.api.app.bsky.graph.getList(
      { list: listUri },
      { headers: sc.getHeaders(dan) },
    )
    expect(forSnapshot(res.data)).toMatchSnapshot()
  })

  it('paginates getList', async () => {
    const full = await agent.api.app.bsky.graph.getList(
      { list: listUri },
      { headers: sc.getHeaders(dan) },
    )
    const first = await agent.api.app.bsky.graph.getList(
      { list: listUri, limit: 1 },
      { headers: sc.getHeaders(dan) },
    )
    const second = await agent.api.app.bsky.graph.getList(
      { list: listUri, cursor: first.data.cursor },
      { headers: sc.getHeaders(dan) },
    )
    const combined = [...first.data.items, ...second.data.items]
    expect(combined).toEqual(full.data.items)
  })

  let otherListUri: string

  it('returns lists associated with a user', async () => {
    const listRes = await agent.api.app.bsky.graph.list.create(
      { repo: alice },
      {
        name: 'new list',
        purpose: 'app.bsky.graph.defs#modlist',
        description: 'blah blah',
        createdAt: new Date().toISOString(),
      },
      sc.getHeaders(alice),
    )
    otherListUri = listRes.uri

    const res = await agent.api.app.bsky.graph.getLists(
      { actor: alice },
      { headers: sc.getHeaders(dan) },
    )
    expect(forSnapshot(res.data)).toMatchSnapshot()
  })

  it('paginates getLists', async () => {
    const full = await agent.api.app.bsky.graph.getLists(
      { actor: alice },
      { headers: sc.getHeaders(dan) },
    )
    const first = await agent.api.app.bsky.graph.getLists(
      { actor: alice, limit: 1 },
      { headers: sc.getHeaders(dan) },
    )
    const second = await agent.api.app.bsky.graph.getLists(
      { actor: alice, cursor: first.data.cursor },
      { headers: sc.getHeaders(dan) },
    )
    const combined = [...first.data.lists, ...second.data.lists]
    expect(combined).toEqual(full.data.lists)
  })

  it('returns a users own list mutes', async () => {
    await agent.api.app.bsky.graph.muteActorList(
      {
        list: otherListUri,
      },
      { encoding: 'application/json', headers: sc.getHeaders(dan) },
    )

    const res = await agent.api.app.bsky.graph.getListMutes(
      {},
      { headers: sc.getHeaders(dan) },
    )
    expect(forSnapshot(res.data)).toMatchSnapshot()
  })

  it('paginates getListMutes', async () => {
    const full = await agent.api.app.bsky.graph.getListMutes(
      {},
      { headers: sc.getHeaders(dan) },
    )
    const first = await agent.api.app.bsky.graph.getListMutes(
      { limit: 1 },
      { headers: sc.getHeaders(dan) },
    )
    const second = await agent.api.app.bsky.graph.getListMutes(
      { cursor: first.data.cursor },
      { headers: sc.getHeaders(dan) },
    )
    const combined = [...first.data.lists, ...second.data.lists]
    expect(combined).toEqual(full.data.lists)
  })

  it('allows unsubscribing from a mute list', async () => {
    await agent.api.app.bsky.graph.unmuteActorList(
      {
        list: otherListUri,
      },
      { encoding: 'application/json', headers: sc.getHeaders(dan) },
    )

    const res = await agent.api.app.bsky.graph.getListMutes(
      {},
      { headers: sc.getHeaders(dan) },
    )
    expect(res.data.lists.length).toBe(1)
  })

  it('updates list', async () => {
    const uri = new AtUri(listUri)
    await agent.api.com.atproto.repo.putRecord(
      {
        repo: uri.hostname,
        collection: uri.collection,
        rkey: uri.rkey,
        record: {
          name: 'updated alice mutes',
          purpose: 'app.bsky.graph.defs#modlist',
          description: 'new descript',
          createdAt: new Date().toISOString(),
        },
      },
      { headers: sc.getHeaders(alice), encoding: 'application/json' },
    )

    const got = await agent.api.app.bsky.graph.getList(
      { list: listUri },
      { headers: sc.getHeaders(alice) },
    )
    expect(got.data.list.name).toBe('updated alice mutes')
    expect(got.data.list.description).toBe('new descript')
    expect(got.data.list.avatar).toBeUndefined()
    expect(got.data.items.length).toBe(2)
  })

  it('embeds lists in posts', async () => {
    const postRef = await sc.post(
      alice,
      'list embed!',
      undefined,
      undefined,
      new RecordRef(listUri, listCid),
    )
    const res = await agent.api.app.bsky.feed.getPosts(
      { uris: [postRef.ref.uriStr] },
      { headers: sc.getHeaders(alice) },
    )
    expect(res.data.posts.length).toBe(1)
    expect(forSnapshot(res.data.posts[0])).toMatchSnapshot()
  })
})
