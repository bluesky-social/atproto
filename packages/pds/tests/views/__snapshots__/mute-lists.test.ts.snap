// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pds views with mutes from mute lists embeds lists in posts 1`] = `
Object {
  "author": Object {
    "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
    "did": "user(0)",
    "displayName": "ali",
    "handle": "alice.test",
    "labels": Array [
      Object {
        "cid": "cids(1)",
        "cts": "1970-01-01T00:00:00.000Z",
        "neg": false,
        "src": "user(0)",
        "uri": "record(1)",
        "val": "self-label-a",
      },
      Object {
        "cid": "cids(1)",
        "cts": "1970-01-01T00:00:00.000Z",
        "neg": false,
        "src": "user(0)",
        "uri": "record(1)",
        "val": "self-label-b",
      },
    ],
    "viewer": Object {
      "blockedBy": false,
      "muted": false,
    },
  },
  "cid": "cids(0)",
  "embed": Object {
    "$type": "app.bsky.embed.record#view",
    "record": Object {
      "$type": "app.bsky.graph.defs#listView",
      "cid": "cids(3)",
      "creator": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "viewer": Object {
          "blockedBy": false,
          "muted": false,
        },
      },
      "description": "new descript",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "updated alice mutes",
      "purpose": "app.bsky.graph.defs#modlist",
      "uri": "record(2)",
      "viewer": Object {
        "muted": false,
      },
    },
  },
  "indexedAt": "1970-01-01T00:00:00.000Z",
  "labels": Array [],
  "likeCount": 0,
  "record": Object {
    "$type": "app.bsky.feed.post",
    "createdAt": "1970-01-01T00:00:00.000Z",
    "embed": Object {
      "$type": "app.bsky.embed.record",
      "record": Object {
        "cid": "cids(2)",
        "uri": "record(2)",
      },
    },
    "text": "list embed!",
  },
  "replyCount": 0,
  "repostCount": 0,
  "uri": "record(0)",
  "viewer": Object {},
}
`;

exports[`pds views with mutes from mute lists flags mutes in threads 1`] = `
Object {
  "$type": "app.bsky.feed.defs#threadViewPost",
  "post": Object {
    "author": Object {
      "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
      "did": "user(0)",
      "displayName": "ali",
      "handle": "alice.test",
      "labels": Array [
        Object {
          "cid": "cids(1)",
          "cts": "1970-01-01T00:00:00.000Z",
          "neg": false,
          "src": "user(0)",
          "uri": "record(2)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(1)",
          "cts": "1970-01-01T00:00:00.000Z",
          "neg": false,
          "src": "user(0)",
          "uri": "record(2)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "followedBy": "record(1)",
        "muted": false,
      },
    },
    "cid": "cids(0)",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "labels": Array [],
    "likeCount": 3,
    "record": Object {
      "$type": "app.bsky.feed.post",
      "createdAt": "1970-01-01T00:00:00.000000Z",
      "text": "again",
    },
    "replyCount": 2,
    "repostCount": 1,
    "uri": "record(0)",
    "viewer": Object {
      "like": "record(4)",
      "repost": "record(3)",
    },
  },
  "replies": Array [
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "did": "user(1)",
          "handle": "carol.test",
          "labels": Array [],
          "viewer": Object {
            "blockedBy": false,
            "followedBy": "record(8)",
            "following": "record(7)",
            "muted": true,
            "mutedByList": Object {
              "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
              "cid": "cids(3)",
              "indexedAt": "1970-01-01T00:00:00.000Z",
              "name": "alice mutes",
              "purpose": "app.bsky.graph.defs#modlist",
              "uri": "record(6)",
              "viewer": Object {
                "muted": true,
              },
            },
          },
        },
        "cid": "cids(2)",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [],
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "of course",
        },
        "replyCount": 0,
        "repostCount": 0,
        "uri": "record(5)",
        "viewer": Object {},
      },
    },
    Object {
      "$type": "app.bsky.feed.defs#threadViewPost",
      "post": Object {
        "author": Object {
          "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
          "did": "user(2)",
          "displayName": "bobby",
          "handle": "bob.test",
          "labels": Array [
            Object {
              "cid": "cids(5)",
              "cts": "1970-01-01T00:00:00.000Z",
              "neg": false,
              "src": "user(2)",
              "uri": "record(11)",
              "val": "self-label-a",
            },
            Object {
              "cid": "cids(5)",
              "cts": "1970-01-01T00:00:00.000Z",
              "neg": false,
              "src": "user(2)",
              "uri": "record(11)",
              "val": "self-label-b",
            },
          ],
          "viewer": Object {
            "blockedBy": false,
            "following": "record(10)",
            "muted": true,
            "mutedByList": Object {
              "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
              "cid": "cids(3)",
              "indexedAt": "1970-01-01T00:00:00.000Z",
              "name": "alice mutes",
              "purpose": "app.bsky.graph.defs#modlist",
              "uri": "record(6)",
              "viewer": Object {
                "muted": true,
              },
            },
          },
        },
        "cid": "cids(4)",
        "embed": Object {
          "$type": "app.bsky.embed.images#view",
          "images": Array [
            Object {
              "alt": "tests/image/fixtures/key-landscape-small.jpg",
              "fullsize": "https://pds.public.url/image/AiDXkxVbgBksxb1nfiRn1m6S4K8_mee6o8r-UGLNzOM/rs:fit:2000:2000:1:0/plain/bafkreigy5p3xxceipk2o6nqtnugpft26ol6yleqhboqziino7axvdngtci@jpeg",
              "thumb": "https://pds.public.url/image/uc7FGfiGv0mMqmk9XiqHXrIhNymLHaex7Ge8nEhmXqo/rs:fit:1000:1000:1:0/plain/bafkreigy5p3xxceipk2o6nqtnugpft26ol6yleqhboqziino7axvdngtci@jpeg",
            },
          ],
        },
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [
          Object {
            "cid": "cids(4)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "did:example:labeler",
            "uri": "record(9)",
            "val": "test-label",
          },
          Object {
            "cid": "cids(4)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "did:example:labeler",
            "uri": "record(9)",
            "val": "test-label-2",
          },
        ],
        "likeCount": 0,
        "record": Object {
          "$type": "app.bsky.feed.post",
          "createdAt": "1970-01-01T00:00:00.000Z",
          "embed": Object {
            "$type": "app.bsky.embed.images",
            "images": Array [
              Object {
                "alt": "tests/image/fixtures/key-landscape-small.jpg",
                "image": Object {
                  "$type": "blob",
                  "mimeType": "image/jpeg",
                  "ref": Object {
                    "$link": "cids(6)",
                  },
                  "size": 4114,
                },
              },
            ],
          },
          "reply": Object {
            "parent": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
            "root": Object {
              "cid": "cids(0)",
              "uri": "record(0)",
            },
          },
          "text": "hear that label_me label_me_2",
        },
        "replyCount": 1,
        "repostCount": 0,
        "uri": "record(9)",
        "viewer": Object {},
      },
    },
  ],
}
`;

exports[`pds views with mutes from mute lists returns a users own list mutes 1`] = `
Object {
  "cursor": "0000000000000::bafycid",
  "lists": Array [
    Object {
      "cid": "cids(0)",
      "creator": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "description": "its me!",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": false,
        },
      },
      "description": "blah blah",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "new list",
      "purpose": "app.bsky.graph.defs#modlist",
      "uri": "record(0)",
      "viewer": Object {
        "muted": true,
      },
    },
    Object {
      "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
      "cid": "cids(2)",
      "creator": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "description": "its me!",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": false,
        },
      },
      "description": "big list of mutes",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice mutes",
      "purpose": "app.bsky.graph.defs#modlist",
      "uri": "record(3)",
      "viewer": Object {
        "muted": true,
      },
    },
  ],
}
`;

exports[`pds views with mutes from mute lists returns lists associated with a user 1`] = `
Object {
  "cursor": "0000000000000::bafycid",
  "lists": Array [
    Object {
      "cid": "cids(0)",
      "creator": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "description": "its me!",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": false,
        },
      },
      "description": "blah blah",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "new list",
      "purpose": "app.bsky.graph.defs#modlist",
      "uri": "record(0)",
      "viewer": Object {
        "muted": false,
      },
    },
    Object {
      "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
      "cid": "cids(2)",
      "creator": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "description": "its me!",
        "did": "user(0)",
        "displayName": "ali",
        "handle": "alice.test",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(0)",
            "uri": "record(2)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": false,
        },
      },
      "description": "big list of mutes",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "name": "alice mutes",
      "purpose": "app.bsky.graph.defs#modlist",
      "uri": "record(3)",
      "viewer": Object {
        "muted": true,
      },
    },
  ],
}
`;

exports[`pds views with mutes from mute lists returns the contents of a list 1`] = `
Object {
  "cursor": "0000000000000::bafycid",
  "items": Array [
    Object {
      "subject": Object {
        "did": "user(0)",
        "handle": "carol.test",
        "labels": Array [],
        "viewer": Object {
          "blockedBy": false,
          "followedBy": "record(1)",
          "muted": true,
          "mutedByList": Object {
            "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
            "cid": "cids(0)",
            "indexedAt": "1970-01-01T00:00:00.000Z",
            "name": "alice mutes",
            "purpose": "app.bsky.graph.defs#modlist",
            "uri": "record(0)",
            "viewer": Object {
              "muted": true,
            },
          },
        },
      },
    },
    Object {
      "subject": Object {
        "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
        "description": "hi im bob label_me",
        "did": "user(1)",
        "displayName": "bobby",
        "handle": "bob.test",
        "indexedAt": "1970-01-01T00:00:00.000Z",
        "labels": Array [
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(1)",
            "uri": "record(3)",
            "val": "self-label-a",
          },
          Object {
            "cid": "cids(1)",
            "cts": "1970-01-01T00:00:00.000Z",
            "neg": false,
            "src": "user(1)",
            "uri": "record(3)",
            "val": "self-label-b",
          },
        ],
        "viewer": Object {
          "blockedBy": false,
          "following": "record(2)",
          "muted": true,
          "mutedByList": Object {
            "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
            "cid": "cids(0)",
            "indexedAt": "1970-01-01T00:00:00.000Z",
            "name": "alice mutes",
            "purpose": "app.bsky.graph.defs#modlist",
            "uri": "record(0)",
            "viewer": Object {
              "muted": true,
            },
          },
        },
      },
    },
  ],
  "list": Object {
    "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
    "cid": "cids(0)",
    "creator": Object {
      "avatar": "https://pds.public.url/image/KzkHFsMRQ6oAKCHCRKFA1H-rDdc7VOtvEVpUJ82TwyQ/rs:fill:1000:1000:1:0/plain/bafkreiaivizp4xldojmmpuzmiu75cmea7nq56dnntnuhzhsjcb63aou5ei@jpeg",
      "description": "its me!",
      "did": "user(2)",
      "displayName": "ali",
      "handle": "alice.test",
      "indexedAt": "1970-01-01T00:00:00.000Z",
      "labels": Array [
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "neg": false,
          "src": "user(2)",
          "uri": "record(5)",
          "val": "self-label-a",
        },
        Object {
          "cid": "cids(2)",
          "cts": "1970-01-01T00:00:00.000Z",
          "neg": false,
          "src": "user(2)",
          "uri": "record(5)",
          "val": "self-label-b",
        },
      ],
      "viewer": Object {
        "blockedBy": false,
        "followedBy": "record(4)",
        "muted": false,
      },
    },
    "description": "big list of mutes",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "name": "alice mutes",
    "purpose": "app.bsky.graph.defs#modlist",
    "uri": "record(0)",
    "viewer": Object {
      "muted": true,
    },
  },
}
`;
